Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -23,142 +20,322 @@ export class FtpOptions {
-            await ftpHelper.uploadFiles(files);
-            uploadSuccessful = true;
-            console.log(tl.loc('UploadSucceedMsg', ftpHelper.progressTracking.getSuccessStatusMessage()));
-
-            tl.setResult(tl.TaskResult.Succeeded, tl.loc('UploadSucceedRes'));
-        } catch (err) {
-            failTask(err);
-        } finally {
-            console.log(tl.loc('DisconnectHost', ftpOptions.serverEndpointUrl.host));
-            ftpClient.end();
-            ftpClient.destroy();
+    let allFiles = tl.find(ftpOptions.rootFolder);
+
+    // filePatterns is a multiline input containing glob patterns
+    tl.debug(
+        "searching for files using: " +
+        ftpOptions.filePatterns.length +
+        " filePatterns: " +
+        ftpOptions.filePatterns
+    );
+
+    // minimatch options
+    let matchOptions = { matchBase: true, dot: true };
+    let win = tl.osType().match(/^Win/);
+    tl.debug("win: " + win);
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug("Candidates found for match: " + allFiles.length);
+    for (let i = 0; i < allFiles.length; i++) {
+        tl.debug("file: " + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    let matchingFilesSet: Set<string> = new Set();
+
+    for (let i = 0; i < ftpOptions.filePatterns.length; i++) {
+        let normalizedPattern: string = path.join(
+            ftpOptions.rootFolder,
+            path.normalize(ftpOptions.filePatterns[i])
+        );
+
+        tl.debug("searching for files, pattern: " + normalizedPattern);
+
+        let matched = tl.match(allFiles, normalizedPattern, null, matchOptions);
+        tl.debug("Found total matches: " + matched.length);
+        // ensure each result is only added once
+        for (let j = 0; j < matched.length; j++) {
+            let match = path.normalize(matched[j]);
+            let stats = tl.stats(match);
+            if (!ftpOptions.preservePaths && stats.isDirectory()) {
+                // if not preserving paths, skip all directories
+            } else if (matchingFilesSet.add(match)) {
+                tl.debug(
+                    "adding " + (stats.isFile() ? "file:   " : "folder: ") + match
+                );
+                if (stats.isFile() && ftpOptions.preservePaths) {
+                    // if preservePaths, make sure the parent directory is also included
+                    let parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug("adding folder: " + parent);
+                    }
+                }
+            }
         }
-    });
+    }
+    return Array.from(matchingFilesSet).sort();
+}
 
-    ftpClient.on('close', (hadErr: boolean) => {
-        console.log(tl.loc('Disconnected'));
-        tl.debug('ftp client close, hadErr:' + hadErr);
-    });
+async function run() {
+    tl.setResourcePath(path.join(__dirname, "task.json"));
+
+    const tries = 3;
```

Comment:
This as a named constant would be nice.

Reply:
I am going to leave this as it is for now.  This may change if we get requests from customer - we may need to make this configurable.

Reply:
👍